src/test/scala/chiselTests/Vec.scala

```diff
@@ -114,6 +115,88 @@ class FillTester(n: Int, value: Int) extends BasicTester {
   stop()
 }
 
+object VecMultiDimTester { 
+  
+  @tailrec
+  private def assert2DIsCorrect(n: Int, arr: Vec[Vec[UInt]], compArr: Seq[Seq[Int]]): Unit = {
+    val compareRow = arr(n) zip compArr(n)
+    compareRow.foreach (x => assert(x._1 === x._2.U))
+    if (n != 0) assert2DIsCorrect(n-1, arr, compArr)
+  }
+
+  @tailrec
+  private def assert3DIsCorrect(n: Int, m: Int, arr: Vec[Vec[Vec[UInt]]], compArr: Seq[Seq[Seq[Int]]]): Unit = {
+    assert2DIsCorrect(m-1, arr(n), compArr(n))
+    if (n != 0) assert3DIsCorrect(n-1, m, arr, compArr)
+  }
+
+  class TabulateTester2D(n: Int, m: Int) extends BasicTester {
+    def gen(x: Int, y: Int): UInt = (x+y).asUInt
+    def genCompVec(x: Int, y:Int): Int = x+y
+    val vec = VecInit.tabulate(n, m){ gen }
+    val compArr = Seq.tabulate(n,m){ genCompVec }
+    
+    assert2DIsCorrect(n-1, vec, compArr)
+    stop()
+  }
+
+  class TabulateTester3D(n: Int, m: Int, p: Int) extends BasicTester {
+    def gen(x: Int, y: Int, z: Int): UInt = (x+y+z).asUInt
+    def genCompVec(x: Int, y:Int, z: Int): Int = x+y+z
+    val vec = VecInit.tabulate(n, m, p){ gen }
+    val compArr = Seq.tabulate(n, m, p){ genCompVec }
+
+    assert3DIsCorrect(n-1, m, vec, compArr)
+    stop()
+  }
+
+  class Fill2DTester(n: Int, m: Int, value: Int) extends BasicTester {
+    val u = VecInit.fill(n,m)(value.U)
+    val compareArr = Seq.fill(n,m)(value)
+    
+    assert2DIsCorrect(n-1, u, compareArr)
+    stop()
+  }
+
+  class Fill3DTester(n: Int, m: Int, p: Int, value: Int) extends BasicTester {
+    val u = VecInit.fill(n,m,p)(value.U)
+    val compareArr = Seq.fill(n,m,p)(value)
+
+    assert3DIsCorrect(n-1, m, u, compareArr)
+    stop()
+  }
+
+  class BidirectionalTester2D(n: Int, m: Int) extends BasicTester {
+    val mod = Module(new PassthroughModule)
+    val vec2D = VecInit.fill(n, m)(mod.io)
+    val tester = Module(new PassthroughModuleTester)
+    for {
+        vec1D <- vec2D
+        module <- vec1D
+    } yield {
+        module <> tester.io
+        assert(module.out === 123.U)
+    }
+    stop()
+  }
+  
+  class BidirectionalTester3D(n: Int, m: Int, p: Int) extends BasicTester {
+    val mod = Module(new PassthroughModule)
+    val vec3D = VecInit.fill(n, m, p)(mod.io)
+    val tester = Module(new PassthroughModuleTester)
+    
+    for {
+        vec2D <- vec3D
+        vec1D <- vec2D
+        module <- vec1D
+    } yield {
+        module <> tester.io
+        assert(module.out === 123.U)
+    }
+    stop()
+  }
+}
```

Comment:
Also there are no tests for the `.tabulate`. I'd do something similar but it might be useful to factor in the indices in the tabulate into the test somehow.

Reply:
The fill's call tabulate, bc they are a special case of tabulate, so I figured by calling 2D/3D fill I would be catching tabulate within that.

Reply:
They do right now, but part of the goal of tests is catching if someone were to refactor things and get it slightly wrong. Perhaps after a refactor `fill` won't call `tabulate` anymore and I'll bet the person doing the refactor won't bother to check the tests.